Initial restructuring of config.py

mfherbst opened the pull request:
This PR will add a parser for environment variables into `config.py`. The intended functionality is:
- [ ] Transparently merge config options from various sources, i.e.
    - Multiple files
    - Environment variables
- [ ] Config data file with metadata about config options
- [ ] Automatically generate documentation from metadata
- [ ] Automatically generate config parser from metadata
- [ ] Normalisation and verification stage done based on the config data

andrenarchy commented:
Phew, 240 LOCs is quite a lot for handling the configuration. In fact, it's almost as much code as the rest of the project together. :sweat_smile: And there are no unit tests yet so it's a bit hard to review. I really like the idea of a config declaration/metadata though!

@mfherbst Do you see a way to simplify this? For example, would a flat config make things easier? At PaperHive we started with nested config dicts and in the end migrated to a flat (one-level) dict because it made many things so much easier. Do you think it'd make any difference here?

mfherbst commented:
Thanks for your comment. Yes, I think a one-level config could make the handling easier. Some other projects use `config["top.middle.low"]` as a key instead of a tree of nested dicts, i.e. `config["top"]["middle"]["low"]` That could help, too. 

On the other hand the user-level syntax is more convenient if a nested dict is used in yaml in my opinion. Contrast
```yaml
storage:
  fulltext: "dir"
  metadata:
     database: "postgres"
     user:     "dbuser"
     password: "pass"
```
to
```yaml
storage.fulltext: "dir"
storage.metadata.database: "postgres"
storage.metadata.user:     "dbuser"
storage.metadata.password: "pass"
```
or alike. Furthermore on the code side it allows to *zoom in* on a part of the config tree.
I thought about a hybrid approach, where the user would supply it in the form mentioned first, but the handling internally would be in the form mentioned last. I did not want to do that because than it might be a bit too inconsistent.

Certainly there should be tests added ;). I just wanted someone's feedback before doing that.

andrenarchy commented:
Yeah, I can see that it would be a bit inconsistent if we have both versions (nested in the file and flat internally and for env vars).

At PaperHive the argument for a flat structure was that env vars *are* a flat structure. Also, we didn't really need the *zoom in* (actually we tried it but it was too inflexible in the end).

My proposal:

```yaml
# config.yml
storage_fulltext: "dir"
storage_metadata_database: "postgres"
storage_metadata_user: "dbuser"
network_parallel_requests: 5
```

```bash
# env vars
STORAGE_FULLTEXT=dir
STORAGE_METADATA_DATABASE=postgres
STORAGE_METADATA_USER=dbuser
NETWORK_PARALLEL_REQUESTS=5
```

We'd then access stuff via `config['storage_fulltext']` in the code – env vars need to be translated to lowercase and converted/sanitized (e.g., into an integer). I found it incredibly useful to formulate the config metadata as functions in code (see [envfefe](https://github.com/paperhive/envfefe#usage), a tiny environment variables parser I wrote for nodejs).

What do you think?

mfherbst commented:
Not a bad idea in fact. And the *zoom in* could be made with a tiny little bit of class code inheriting from dict. I like that. I don't like the idea that metadata are functions. I think they really should be data. But a mapping `"Type" -> "Parser function"` is definitely sensible. I am happy to make those changes.

From tomorrow my working speed will be a little slower, since I will have less time to work on `libfulltext`. Just that you are not surprised ;).

andrenarchy commented:
@mfherbst do we actually need zooming in right now? If not we can also add it later if/when it makes sense. And function vs data: up to you :) (I just liked it because the config is coupled to the code anyway, it was shorter and I got proper type information via TypeScript).

mfherbst commented:
We had it at some point afaik. We don't use it now, though. Not sure it makes sense at the moment. I agree we should add it later as needed.

mfherbst commented:
@andrenarchy I implemented what we discussed, namely flattened the configuration file and added some unit tests. What is missing is the parser for the environment variables if their string value is not sufficient. Feel free to take a look and drop me a line if you have further comments.

mfherbst commented:
All right now I got the parsers as well including tests.

mfherbst commented:
@andrenarchy ping

andrenarchy commented:
I'm not really sure how we want to proceed with this PR. It seems a bit extreme to include >500 LOCs for setting only a couple of options (right now we only have 2 config options!). Is there any easy way to simplify this? The PR weighs as much as the rest of the code at the moment and it'd be great to reduce complexity? 

@mfherbst @pseyfert What's your opinion? I'm sorry that I'm a bit picky with this – I just think we should talk about it. :) 

Perhaps this is even a nice thing to publish as a separate lib? I didn't do much research about other libs so maybe you can share your thoughts on this.

mfherbst commented:
Rebased it.

Discussion is great. No worries. I really tend to make things too general and complicated too quickly ... and this way you can hold me back.

Regards the **2 config options**: It's not going to stay like that and as far as I see it with this PR merged we never really have to worry about config handling any more. I am not aware of a good config handling library in python, but I am happy to see options if there are. The way I do it here is inspired from [qutebrowser](https://github.com/qutebrowser/qutebrowser) by the way. I really enjoyed that from a user perspective.

**Simplification**: I tried, but perhaps there is more. You always overlook stuff if you are in your own mind structure. Let me know if you have ideas.

andrenarchy commented:
@mfherbst What do you think about having it as a separate project? It might be useful for others, don't you think?
